compiler-core/src/type_/tests.rs

```diff
@@ -2142,6 +2142,28 @@ pub fn fun(x) {
     );
 }
 
+#[test]
+// https://github.com/gleam-lang/gleam/issues/3797
+fn type_unification_removes_inferred_variant_in_tuples() {
+    assert_module_error!(
+        r#"
+type Either(a, b) {
+  Left(value: a)
+  Right(value: b)
+}
+
+fn a_or_b(_first: value, second: value) -> value {
+  second
+}
+
+pub fn main() {
+  let #(right) = a_or_b(#(Left(5)), #(Right("hello")))
+  Left(..right, inner: 10)
+}
+"#
+    );
+}
```

Comment:
Can we have a test for in a function too? Just to ease my mind!

Reply:
I'm not exactly sure what you mean by that

Reply:
You written a test to ensure this problem doesn't happen for tuples, but I can't see a test that shows the same wouldn't happen with anonymous functions.

Reply:
Ah I see thanks, I'll add that shortly